src/server/pkg/sync/sync.go

```diff
@@ -4,16 +4,31 @@ package sync
 import (
 	"os"
 	"path/filepath"
+	"sync"
 	"syscall"
 
 	pachclient "github.com/pachyderm/pachyderm/src/client"
 	"github.com/pachyderm/pachyderm/src/client/pfs"
 	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
 
-	log "github.com/Sirupsen/logrus"
 	"golang.org/x/sync/errgroup"
 )
 
+// Puller as a struct for managing a Pull operation.
+type Puller struct {
```

Comment:
3 minor suggestions:

1) you could write the struct like this
```
type Puller struct {
  sync.Mutex
  pipes map[string]bool
  errCh chan error
}
```
then later just do:
```
p := NewPuller()
p.Lock()
```

2) Would you mind putting a comment on pipes? (e.g. // pipes[<path>] == true iff <path> was never read and a goroutine is blocked on it.)

3) Would it make sense for pipes to be `map[string]struct{}`? It looks like you're just using it as a set. It's a little more awkward to set values, though, so I really don't feel strongly about this